multicluster/controllers/multicluster/leader/stale_controller.go

```diff
@@ -0,0 +1,234 @@
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if err == nil && memberAnnounce.DeletionTimestamp.IsZero() {
+		// Ignore the event if it's not with non-zero DeletionTimestamp
+		return ctrl.Result{}, nil
+	}
+
+	// Clean up all corresponding ResourceExports when the member cluster's
+	// MemberClusterAnnounce is deleted.
+	clusterID := getClusterIDFromName(req.Name)
+	staleResExports, err := getResourceExportsByClusterIDFunc(c, ctx, clusterID)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get ResourceExports by ClusterID", "clusterID", clusterID)
+		return ctrl.Result{}, err
+	}
+	if !deleteResourceExports(ctx, c.Client, staleResExports) {
+		return ctrl.Result{}, fmt.Errorf("failed to clean up all stale ResourceExports for the member cluster %s, retry later", clusterID)
+	}
+
+	// When cleanup is done, remove the Finalizer of this MemberClusterAnnounce.
+	finalizer := fmt.Sprintf("%s/%s", MemberClusterAnnounceFinalizer, memberAnnounce.ClusterID)
+	memberAnnounce.Finalizers = common.RemoveStringFromSlice(memberAnnounce.Finalizers, finalizer)
+	if err := c.Update(context.TODO(), memberAnnounce); err != nil {
+		klog.ErrorS(err, "Failed to update MemberClusterAnnounce", "MemberClusterAnnounce", klog.KObj(memberAnnounce))
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (c *StaleResCleanupController) SetupWithManager(mgr ctrl.Manager) error {
+	// Add an Indexer for ResourceExport, so it can be filtered by the ClusterID.
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &mcv1alpha1.ResourceExport{}, indexKey, func(rawObj client.Object) []string {
+		resExport := rawObj.(*mcv1alpha1.ResourceExport)
+		return []string{resExport.Spec.ClusterID}
+	}); err != nil {
+		klog.ErrorS(err, "Failed to create the index")
+		return err
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&mcv1alpha1.MemberClusterAnnounce{}).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: 1,
+		}).
+		Complete(c)
+}
+
+func cleanUpStaleResourceExports(ctx context.Context, mgrClient client.Client) error {
+	existingMemberClusterIDs, err := getExistingMemberClusterIDs(ctx, mgrClient)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get existing member cluster's ClusterID")
+		return err
+	}
+
+	resourceExports := &mcv1alpha1.ResourceExportList{}
+	err = mgrClient.List(ctx, resourceExports, &client.ListOptions{})
+	if err != nil {
+		klog.ErrorS(err, "Failed to get ResourceExports")
+		return err
+	}
+
+	staleResExports := []mcv1alpha1.ResourceExport{}
+	for _, resExport := range resourceExports.Items {
+		// The AntreaClusterNetworkPolicy kind of ResourceExport is created in the leader directly
+		// without a ClusterID info. It's not owned by any member cluster.
+		if resExport.Spec.Kind != constants.AntreaClusterNetworkPolicyKind && !existingMemberClusterIDs.Has(resExport.Spec.ClusterID) {
```

Comment:
Not familiar with the code, but should the first expression use `==`? Otherwise non AntreaClusterNetworkPolicy will be collected by staleResExports.

Reply:
Yes, we did this to skip AntreaClusterNetworkPolicy ResourceExport because these kind of ResourceExport are created by user, we don't want to clean up them automatically.